### PR TITLE
Implement iterators task #2 - ENikonov

### PR DESCRIPTION
**Pull request by re-nikonov**

### CODE REVIEW

File: iterators task (2)/student/include/impl.h [hunk 1]
```diff
@@ -1,33 +1,53 @@
...
 template<class Iterator, class T>
 struct pair_iterator<Iterator, false, T> : public Iterator
 {
-    /** @todo Iterator traits */
-    /** @todo Iterator operations */
-    /** @todo Constructor from original iterator */
+    using value_type = typename T::second_type;
+    using reference =
+        typename std::conditional<
+            std::is_const<typename std::remove_reference<typename std::iterator_traits<Iterator>::reference>::type>::value,
```
Comment (smay1613) [inline iterators task (2)/student/include/impl.h:39]: Nice!
Consider is_const trait extraction to a named struct

File: iterators task (2)/student/include/impl.h [hunk 2]
```diff
@@ -1,33 +1,53 @@
...
+            const value_type&,
+            value_type&
+        >::type;
+    using pointer = typename std::add_pointer<reference>::type;
+
+    explicit pair_iterator(Iterator iterator) : Iterator(std::move(iterator)) {}
```
Comment (smay1613) [inline iterators task (2)/student/include/impl.h:45]: why do you need move here?

Comment (re-nikonov) [reply iterators task (2)/student/include/impl.h:45]: I don't. Just used to move everything what isn't lvalue reference. But this is just an iterator... removed std::move().

File: iterators task (2)/student/include/impl.h (hunk 1 above)
Comment (re-nikonov) [reply iterators task (2)/student/include/impl.h:39]: Done! But the merge request has been closed... Anyway it was interesting!

**Pull request closed without merging**